### PR TITLE
[Issue 2448] use v1 search endpoint and frontend fetch pattern refactors

### LINKED ISSUE #2448: Switch frontend to use v1 Search endpoint
### Summary

Search currently uses a v0.1 endpoint that uses database search. This ticket is to update the frontend to use the v1 API backed by OpenSearch.

🚨🚨🚨 Can't go to main before #1656 is fully ready in Dev and can't be Released until #1656 is in Prod 🚨🚨🚨

### Acceptance criteria

- [ ] Frontend updated to use v1 Search API

### PR DESCRIPTION
**Pull request by doug-s-nava**
## Summary
Fixes #2448

## Changes proposed
**DISCLAIMER**: there is a lot of scope creep in this PR. Feel free to push back on anything you feel goes to far here.

The primary purpose of this change is to change the search API to use the `v1` version of the search endpoint in order to activate the use of Open Search on the backend. This purpose is accomplished in the first commit, which can be split into a separate PR or commit if called for.

The rest of the PR contains refactors to the `API` fetch system that:

* should reduce redundancy and make better use of the classical setup in place here
* makes better use of available TS functionality
* better encapsulates domain specific logic (eg moving search specific functionality out of the base class)
* fixes bugs

### Test Steps

1. visit the search page, and perform all manner of searches with different filters, queries, and sort options
2. _VALIDATE_: everything works as expected
3. visit various opportunity pages
4. _VALIDATE_: everything works as expected

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   "scripts": {
     "build": "next build",
     "dev": "next dev",
+    "debug": "NODE_OPTIONS='--inspect' next dev",
```
Comment (doug-s-nava) [inline frontend/package.json:11]: I was not able to get this to work to use the chrome node debugger, and ended up using the VS code debugger, which did work to debug node based / server side next functions. If anyone else can get this to work, we should probably keep this, or we can just get rid of it.

see https://github.com/vercel/next.js/issues/48767 and https://nextjs.org/docs/pages/building-your-application/configuring/debugging

File: frontend/src/app/api/BaseApi.ts [hunk 2]
```diff
@@ -15,32 +15,26 @@ import {
   ValidationError,
 } from "src/errors";
 import { QueryParamData } from "src/services/search/searchfetcher/SearchFetcher";
-// TODO (#1682): replace search specific references (since this is a generic API file that any
```
Comment (doug-s-nava) [inline frontend/src/app/api/BaseApi.ts:18]: I believe I've accomplished this now, though the ticket mentioned here was already closed.

File: frontend/src/app/api/BaseApi.ts [hunk 3]
```diff
@@ -15,32 +15,26 @@ import {
...
-
 export interface HeadersDict {
   [header: string]: string;
 }
 
 export default abstract class BaseApi {
-  // Root path of API resource without leading slash.
-  abstract get basePath(): string;
+  // Root path of API resource without leading slash, can be overridden by implementing API classes as necessary
+  get basePath() {
```
Comment (doug-s-nava) [inline frontend/src/app/api/BaseApi.ts:31]: since we are using the same basePath everywhere, I figured we may as well define it at a higher level

File: frontend/src/app/api/BaseApi.ts [hunk 4]
```diff
@@ -54,29 +48,28 @@ export default abstract class BaseApi {
...
     }
+    headers["Content-Type"] = "application/json";
     return headers;
   }
 
   /**
    * Send an API request.
    */
-  async request(
+  async request<ResponseType extends APIResponse>(
```
Comment (doug-s-nava) [inline frontend/src/app/api/BaseApi.ts:58]: use a generic to get around having to hardcode search specific typing

File: frontend/src/app/api/BaseApi.ts [hunk 5]
```diff
@@ -149,14 +132,14 @@ export function createRequestUrl(
   let url = [...cleanedPaths].join("/");
   if (method === "GET" && body && !(body instanceof FormData)) {
     // Append query string to URL
-    const body: { [key: string]: string } = {};
+    const newBody: { [key: string]: string } = {};
```
Comment (doug-s-nava) [inline frontend/src/app/api/BaseApi.ts:135]: I think this was completely broken before, but was unused so nobody noticed

File: frontend/src/app/api/SearchOpportunityAPI.ts [hunk 6]
```diff
@@ -1,37 +1,48 @@
...
+  | "agency"
+  | "category";
+
+const filterNameMap = {
+  status: "opportunity_status",
+  fundingInstrument: "funding_instrument",
+  eligibility: "applicant_type",
+  agency: "agency",
+  category: "funding_category",
+} as const;
```
Comment (doug-s-nava) [inline frontend/src/app/api/SearchOpportunityAPI.ts:35]: these types are all to support the refactor of buildFilters, which is... admittedly a lot of scope creep for this ticket

File: frontend/src/app/api/SearchOpportunityAPI.ts [hunk 7]
```diff
@@ -1,37 +1,48 @@
...
-  get headers() {
-    const baseHeaders = super.headers;
-    const searchHeaders = {};
-    return { ...baseHeaders, ...searchHeaders };
-  }
-
   async searchOpportunities(searchInputs: QueryParamData) {
     const { query } = searchInputs;
-    const filters = this.buildFilters(searchInputs);
-    const pagination = this.buildPagination(searchInputs);
```
Comment (doug-s-nava) [inline frontend/src/app/api/SearchOpportunityAPI.ts:34]: these function do not reference the class at all, so can be split out

File: frontend/src/app/api/SearchOpportunityAPI.ts [hunk 8]
```diff
@@ -44,93 +55,68 @@ export default class SearchOpportunityAPI extends BaseApi {
...
       }
     }
+  }
 
-    // default to descending
-    let sort_direction: PaginationSortDirection = "descending";
-    if (sortby) {
-      sort_direction = sortby?.endsWith("Desc") ? "descending" : "ascending";
-    }
+  const sort_direction =
```
Comment (doug-s-nava) [inline frontend/src/app/api/SearchOpportunityAPI.ts:113]: the diff is tough here since I removed these functions from the class, but the only real change I made to the pagination function is here, simplifying the sort_direction logic slightly.

File: frontend/tests/api/SearchOpportunityApi.test.ts [hunk 9]
```diff
@@ -1,113 +1,219 @@
...
+    jest.restoreAllMocks();
+  });
+
+  it("instantiates correctly", () => {
+    expect(searchApi.namespace).toEqual("opportunities");
   });
 
   describe("searchOpportunities", () => {
-    beforeEach(() => {
-      global.fetch = mockFetch({});
```
Comment (doug-s-nava) [inline frontend/tests/api/SearchOpportunityApi.test.ts:46]: this test was mocking down to the fetch level, while the existing opportunity listing API test was just mocking the request function on the class. In the interest of using inheritance to our benefit and keeping the tests somewhat uniform, I changed this to use the request mocking approach, and left the fetch mocking approach for the base class

**Review by acouch — APPROVED** [review]
This cleanup looks great. Out of scope but we could also remove the mock API at some point as I don't think we are using it anywhere and it was never fully built AFAICT.
